packages/controls/src/widget_dom_events.ts

```diff
@@ -0,0 +1,275 @@
+        }
+        console.log('Adding backbone listener for changes in view count')
+        this.listenTo(current_model, 'change:_view_count', this.update_listeners)
+        this.update_listeners()
+    }
+
+    update_listeners() {
+        // Remove all existing DOM event listeners
+        this.remove_listeners()
+        // Add watchers to any existing views of the model
+        this.attach_listeners()
+    }
+
+    remove_listeners() {
+        // Remove all of the event listeners stored in the cache.
+        for (let listener of this.get('_attached_listeners')) {
+            console.log('Removing ', listener.event, ' from ', listener.view, 'function', listener.func)
+            listener.view.el.removeEventListener(listener.event, listener.func)
+        }
+        this.set('_attached_listeners', [])
+    }
+
+    _add_listeners_to_view(view) {
+        for (let event of this.get('watched_events')) {
+            switch (this.key_or_mouse(event)) {
+                case "keyboard":
+                    this._add_key_listener(event, view)
+                    break
+                case "mouse":
+                    let handler = this._dom_click.bind(this, view)
+                    view.el.addEventListener(event, handler)
+                    // Keep track of the listeners we are attaching so that we can
+                    // remove them if needed.
+                    this._cache_listeners(event, view, handler)
+                    break
+                default:
+                    console.log('Not familiar with that message source')
+                    break
+            }
+        }
+    }
+
+    attach_listeners() {
+        // console.log("Attaching listeners")
+        let current_source = this.get('source')
+        _.each(current_source.views, (view_promise) => {
+            Promise.resolve(view_promise).then((view) => {
+                this._add_listeners_to_view(view)
+            })
+        })
+    }
+
+    _add_key_listener(event_type, view) {
+        // Key listeners should:
+        //     + Only fire when the mouse is over the element.
+        //     + Not propagate up to the notebook (because imagine you
+        //       press 'x' on your widget and cut the cell...not what
+        //       you probably want the user to experience)
+        //
+        // The approach here is to add a key listener on mouseenter and remove
+        // it on mouseleave.
+        let key_handler = (event) => {
+            // console.log('Key presses FTW!', event)
+            if (this.get('ignore_modifier_key_events') &&
+               _.contains(this.get('_modifier_keys'), event.key)) {
+                // If the key event is supposed to be ignored, then skip it.
+                return
+            }
+            this._send_dom_event(event)
+            // Need this (and useCapture in the listener) to prevent the keypress
+            // from propagating to the notebook.
+            event.stopPropagation()
+        }
```

Comment:
I think we need a `stopPropagation()` or possibly a `preventDefault()` in the case of button 1 press and drag (if the user is listening for mousedown/mousemotion/mouseup events (otherwise the browser tries to drag the element).  Also for right mouse button mousedown/click, otherwise the browser tries to pop up a context menu.